notifications: Rename `private_message` to `direct_message`.

prakhar1144 opened the pull request:
[CZO message](https://chat.zulip.org/#narrow/stream/2-general/topic/private.20messages.20.3D.3E.20direct.20messages/near/1588736):
> It's probably be a while before we manage to rename everything, but I think we should do any renames that are easy to verify as we notice them. (Some of the backend stuff is tricky, like the notification triggers, because items can be in the API, database, or queues -- **@Prakhar-Pratyush** FYI for the `private_message` notification trigger value that sits in queues, mainly since you're doing a similar rename right now for topic wildcard mentions and it might make sense for you to change them both while it's fresh)

---
* The First commit renames the `PRIVATE_MESSAGE` attribute of the `NotificationTriggers` class to `DIRECT_MESSAGE`. -- custom migration to update the existing value in the database + API changes.
* The Second commit renames the keyword `pm` to `dm` in the `RecipientInfoResult` dataclass. -- handles the fields that sit in the queues.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

---
**Manual Testing**

I did manual testing by adding the translation code (`TODO/compatibility`) in `event_queue.py` and keeping the field in the event sent by `message_send` and `message_edit` codepath as `pm_mention_email_disabled_user_ids` (instead of `dm_mention_email_disabled_user_ids`).

* Disabled the setting "email notification for DM, mentions, and alerts"

* No missed message emails were sent.

* Without the compatibility code i.e. simply using the following code block,
``` 
dm_mention_push_disabled_user_ids = set(event_template.pop("dm_mention_push_disabled_user_ids", []) 
```
missed message emails were sent.

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
@zulipbot add "mentor review"

abhijeetbodas2001 commented on zerver/tornado/event_queue.py:
```diff
@@ -973,12 +991,33 @@ def process_message_event(
+        )
+    elif "pm_mention_email_disabled_user_ids" in event_template:
+        dm_mention_email_disabled_user_ids = set(
+            event_template.get("pm_mention_email_disabled_user_ids", [])
```
If we are doing the `x in y` check earlier, we don't need to provide a default value of `[]` when we do `.get(x)`.

timabbott replied:
Yeah, this would be a nice simplification.

abhijeetbodas2001 replied:
Prakhar mentioned this is PM, the issue here is the `set` function cannot handle an `Optional` argument (which `.get` returns without a provided default)

> zerver/tornado/event_queue.py:1001: error: Argument 1 to "set" has incompatible type "Optional[Any]"; expected "Iterable[Any]" [arg-type]

timabbott replied:
OK.

abhijeetbodas2001 reviewed (commented):
Looks good to me :+1: 

---
This is general thought and not specific to this PR, but one thing I'm worried about is the increasing amount of compatibility code we have accumulated in the `even_queue.py` file.

There are cases when we absolutely can't avoid compatibility code (for example, for large migrations handled by `reformat_legacy_send_message_event`). In other cases, we have compatibility code mainly for correct user experience. The changes in this PR are examples  of such code. Even if we don't have the compatibility code here (and just provide the default values for keys here), the worst that can happen is users receiving a few more/few less notifications after server upgrade. Unlike the "must-have" compatibility code like in `reformat_legacy_send_message_event`, not having which will lead to runtime errors.

Now I'm guessing that a server upgrade is a sufficiently infrequent event that we can skip writing code which falls in the "correct user experience" bucket. @timabbott does that sound correct? I think the only place where server upgrades are much more frequent is in the Zulip-cloud offering.

timabbott commented:
I think the compatibility code is well worth doing -- it's not a huge amount of work, we have a clear story for when we get to delete it (when there's an intermediate upgrade that would happen first), and it means that we don't have to handle "production help" reports from folks who get the types of weird exceptions that would occur if we didn't write that code, which in aggregate can be quite expensive.

timabbott commented on zerver/lib/push_notifications.py:
```diff
@@ -659,12 +659,12 @@ def get_gcm_alert(
     elif (
         message.recipient.type == Recipient.PERSONAL
-        and trigger == NotificationTriggers.PRIVATE_MESSAGE
+        and trigger == NotificationTriggers.DIRECT_MESSAGE
```
Hmm, where is the compatibility code for handling "private_message" in a push notifications queue?

prakhar1144 replied:
Hm, the compatibility code was missing.

Added the compatibility code + did manual testing by changing the trigger value in `get_push_notification_trigger`

prakhar1144 commented:
@timabbott This is ready to review.

Updates in the recent push:
* Addressed the review comments. (I've replied to the respective comments)
* We missed the compatibility code for push notifications while renaming the trigger values in f4fa82e. 
In f4fa82ea2065703e357f365707ea6215e4c3a966, we renamed the following notification triggers:
   * `wildcard_mentioned` to `stream_wildcard_mentioned`
   * `followed_topic_wildcard_mentioned` to `stream_wildcard_mentioned_in_followed_topic`.

    Fixed that in the first commit.

prakhar1144 commented:
@zulipbot add "integration review"

timabbott commented:
Nice, merged, thanks so much for carefully testing those transitions! 

This is the kind of thing where it's a nontrivial amount of work to rename something like this that has some persistent state in the database, but the payoff is pretty significant in the long term in terms of having the code be readable in the future, so we generally err on the side of just figuring out how to do these transitions correctly.
